Issue #7187: AsOf Join Performance

hawkfish opened the pull request:
Rework the implementation to use the partitioning library on the left side.
This lets us perform joins on matching partitions and avoid repeated scans.
It also reduces the run time of the new small probe benchmark by about 200x (16s => 77ms)
so it is only about 2.5x slower than the hand-crafted inner join (30ms).

Commit:
Issue #7187: AsOf Benchmark

Add a benchmark from the discussion.

Commit:
Issue #7187: AsOf Performance

Small refactor of operator code into operator state.

Commit:
Merge branch 'master' into asof-perf

Commit:
Issue #7187: AsOf Performance

Sort lhs payload, not selection.
Prep for multi-chunk buffering.

Commit:
Merge branch 'master' into asof-perf

Commit:
Issue #7187: AsOf Performance

Refactor into smaller methods.

Commit:
Issue #7187: AsOf Performance

Refactor for larger buffers.

Commit:
Merge branch 'master' into asof-perf

Commit:
Merge branch 'feature' into asof-perf

Commit:
Merge branch 'master' into asof-perf

Commit:
Issue #7187: AsOf Performance

Refactor for deferred flushing

Commit:
Issue #7187: AsOf Performance

Convert to buffered sorting.

Commit:
Merge branch 'master' into asof-perf

Commit:
Issue #7187: AsOf Performance

Return impossible matches immediately so simplify the rest of the join logic.

Commit:
Issue #7187: AsOf Performance

Switch to using partitioning on both sides so that scans are minimised.

Commit:
Merge branch 'master' into asof-perf

Commit:
Issue #7187: AsOf Performance

Parallelise left side sorting.

Commit:
Issue #7187: AsOf Performance

Fix stale left outer flag indexing.

Mytherin commented on src/execution/operator/join/physical_asof_join.cpp:
```diff
@@ -392,273 +489,325 @@ void AsOfLocalState::ResolveJoin(DataChunk &input, bool *found_match, std::pair<
-	auto &lstate = lstate_p.Cast<AsOfLocalState>();
-	auto &gsink = sink_state->Cast<AsOfGlobalSinkState>();
+void AsOfProbeBuffer::ResolveSimpleJoin(ExecutionContext &context, DataChunk &chunk) {
+	auto &gsink = op.sink_state->Cast<AsOfGlobalSinkState>();
```
This appears to be uncovered by the tests

Mytherin commented on src/execution/operator/join/physical_asof_join.cpp:
```diff
@@ -270,120 +261,226 @@ void AsOfLocalState::ResolveJoinKeys(DataChunk &input) {
+
+	if (gsink.rhs_sink.count == 0) {
+		// empty RHS
+		if (!EmptyResultIfRHSIsEmpty()) {
```
This appears to be uncovered by the tests

Mytherin commented on src/execution/operator/join/physical_asof_join.cpp:
```diff
@@ -270,120 +261,226 @@ void AsOfLocalState::ResolveJoinKeys(DataChunk &input) {
-		left.SetIndex(i);
+	left_outer.Reset();
+	if (!right_itr) {
+		return;
```
This appears to be uncovered

Mytherin commented on src/execution/operator/join/physical_asof_join.cpp:
```diff
@@ -392,273 +489,325 @@ void AsOfLocalState::ResolveJoin(DataChunk &input, bool *found_match, std::pair<
+				//	Wait for all threads to finish
+				//	TODO: How to implement a spin wait correctly?
+				//	Returning BLOCKED seems to hang the system.
+				std::this_thread::yield();
```
None of the yield calls appear to be covered by the coverage tests

hawkfish replied:
The skew test appears to cover this as well.

Mytherin commented on src/execution/operator/join/physical_asof_join.cpp:
```diff
@@ -392,273 +489,325 @@ void AsOfLocalState::ResolveJoin(DataChunk &input, bool *found_match, std::pair<
+				//	More to flush
+				lsource.probe_buffer.BeginLeftScan(left_bin);
+			} else if (!IsRightOuterJoin(join_type)) {
+				return SourceResultType::FINISHED;
```
Also appears to be uncovered

hawkfish replied:
That was fun...

Mytherin reviewed (commented):
Thanks for the PR! Great results. Some comments below. I think it would be great if we could get some more stress testing/larger ASOF join tests - particularly as this is a quite large rework of how the operator works and we don't have that many ASOF joins on larger data sets currently.

I've annotated some lines that [appear to be uncovered by the coverage tests](https://app.codecov.io/github/duckdb/duckdb/commit/dba6cbbb96e37dcbc49e7b74765219e3f2f1118e/blob/src/execution/operator/join/physical_asof_join.cpp). Perhaps those are a good starting point?

Perhaps another idea for a stress test is to e.g. take TPC-H, increment the join keys in one of the tables by `0.1` (e.g. `l_orderkey` or `o_orderkey`), then do an ASOF join and verify the results are the same.

lnkuiper reviewed (approved):
Awesome performance improvements. I'm excited to see that the parallel state machine pattern is being used here! So much more convenient to use the morsel-driven parallelism to go through different stages than using the task-based parallelism when the operator is a `Source`.

When I wrote the parallel state machine for the hash join, I found that I needed a lot of tests for all of the join types to make the code robust. OUTER/RIGHT especially as this forces the build side to be scanned when the operator is a source, and therefore stresses the parallel state machine more.

I didn't have to write many new tests because I added the `pragma verify_external` setting and added that to a bunch of join tests, perhaps something similar is possible for the ASOF join?

Commit:
Merge branch 'feature' into asof-perf

hawkfish commented on src/execution/operator/join/physical_asof_join.cpp:
```diff
@@ -392,273 +489,325 @@ void AsOfLocalState::ResolveJoin(DataChunk &input, bool *found_match, std::pair<
+	auto &rhs_sink = gsource.gsink.rhs_sink;
+
+	//	Step 1: Combine the partitions
+	lsource.CombineLeftPartitions();
```
I thought about making a double pipeline breaker, but `ExecuteInternal` returns left values that cannot pass the predicates because they are `NULL` so this needs to be done in the source.

Commit:
Issue #7187: AsOf Performance

Code coverage and a few fixes they uncovered.

Commit:
Issue #7187: AsOf Performance

Force external just to be sure.

hawkfish commented:
> I didn't have to write many new tests because I added the `pragma verify_external` setting and added that to a bunch of join tests, perhaps something similar is possible for the ASOF join?

Thanks - I added it to the highly skewed test.

Commit:
Issue #7187: AsOf Performance

Add TPC-H join comparison test.

Commit:
Merge branch 'feature' into asof-perf

Commit:
Merge branch 'master' into asof-perf

Commit:
Merge branch 'master' into asof-perf

Commit:
Merge branch 'feature' into asof-perf

Commit:
Merge branch 'feature' into asof-perf

Mytherin commented:
Thanks - looks like the code coverage test is still failing. Could we increment the uncovered lines or add tests that cause these lines to be covered - in particular the `partition_state.cpp` and `radix_partitioning.cpp`?

Note `test_coverage` are currently not run for the coverage due to CI time constraints (the fast tests already take almost 3 hours).

hawkfish commented:
> Note `test_coverage` are currently not run for the coverage due to CI time constraints (the fast tests already take almost 3 hours).

So try to make some relative ly quick tests that are either `.test` or `.test_slow`?

Mytherin commented:
Only `.test` is run for the coverage currently

Commit:
Merge branch 'feature' into asof-perf

hawkfish commented:
I can't make fast tests that stress this part of the code. Basically the merging and repartitioning only happened with large data sets and memory pressure.

Commit:
Issue #7187: AsOf Performance Coverage

Remove dead code.
Tweak uncovered row count for new formatting.

hawkfish commented:
Had some thoughts on this so may give it a go.
